src/index/scanners/default.rs

```diff
@@ -408,3 +324,81 @@ impl SearchBuilder for DefaultBuilder {
         }))
     }
 }
+
+type Extra<'b> = &'b mut (NonZero<u64>, u16, &'b mut [u32]);
+type Result<'b> = ((Reverse<Distance>, AlwaysEqual<()>), AlwaysEqual<Extra<'b>>);
+
+#[inline(always)]
+fn rerank_index_wrapper<'a, O: Operator>(
+    vector: O::Vector,
+    relation: &'a (impl RelationPrefetch + RelationReadStream),
+    mut fetcher: impl SearchFetcher + 'a,
+    opfamily: Opfamily,
+    results: Vec<Result<'a>>,
+    io_rerank: SearchIo,
+) -> Box<dyn Iterator<Item = (f32, NonZero<u64>)> + 'a> {
+    let prefilter = move |payload| {
+        if !prererank_filtering() {
+            return true;
+        }
+        let (key, _) = pointer_to_kv(payload);
+        fetcher.filter(key)
+    };
+    match io_rerank {
+        SearchIo::ReadBuffer => {
+            let prefetcher = PlainPrefetcher::<_, BinaryHeap<_>>::new(relation.clone(), results);
+            Box::new(
+                rerank_index::<O, _, _>(vector, prefetcher, prefilter)
+                    .map(move |(distance, payload)| (opfamily.output(distance), payload)),
+            )
+        }
+        SearchIo::PrefetchBuffer => {
+            let prefetcher = SimplePrefetcher::new(relation.clone(), results);
+            Box::new(
+                rerank_index::<O, _, _>(vector, prefetcher, prefilter)
+                    .map(move |(distance, payload)| (opfamily.output(distance), payload)),
+            )
+        }
+        SearchIo::ReadStream => {
+            let prefetcher = StreamPrefetcher::new(relation, results);
+            Box::new(
+                rerank_index::<O, _, _>(vector, prefetcher, prefilter)
+                    .map(move |(distance, payload)| (opfamily.output(distance), payload)),
+            )
+        }
+    }
+}
+
+#[inline(always)]
+fn rerank_heap_wrapper<'a, O: Operator>(
+    vector: O::Vector,
```

Comment:
don't handle `rerank_heap` and `rerank_index` differentl; split `algorithm::insert` into two functions, `algorithm::insert_vector` and `algorithm::insert_index`, so that the full-precision vectors in the index are unprojected, too